### PR TITLE
Gh-1345 Extract Items

### PR DESCRIPTION
**Pull request by m607123**
Operations now called Map and FlatMap - these operations take in an Iterable, or an Iterable of Iterables respectively, returning an object or an iterable respectively.
This is achieved by executing a function on the input (see gchq/koryphe#60).

Should not be merged until the Koryphe branch above is completed and merged in.

### CODE REVIEW

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/ExtractItemsHandler.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import uk.gov.gchq.gaffer.operation.OperationException;
+import uk.gov.gchq.gaffer.operation.impl.ExtractItems;
+import uk.gov.gchq.gaffer.store.Context;
+import uk.gov.gchq.gaffer.store.Store;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ExtractItemsHandler implements OutputOperationHandler<ExtractItems, Iterable<? extends Object>> {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/ExtractItemsHandler.java]: Could you add a comment here please?

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/ExtractItemsHandler.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public List<? extends Object> doOperation(final ExtractItems operation, final Context context, final Store store) throws OperationException {
+        if (null == operation) {
+            throw new OperationException("Operation cannot be null");
+        }
+
+        final Iterable<Iterable<?>> input = operation.getInput();
+
+        if (null == input) {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/ExtractItemsHandler.java]: What will happen if the selection value is greater than the size of the Iterables contained in the input? Should we try to protect against this? Ideally this validation would take place in the ```Operation#validate()``` method.

**Comment by p013570** [comment]
After our discussion ExtractItems should be converted to a Map operation which takes a function to map the iterable input. We will also need a NthItem function to allow us to extract the nth item.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * @param <I_ITEM> The object type of the input iterable
+ * @param <O_ITEM> The object type of the output object
+ */
+public class Map<I_ITEM, O_ITEM> implements
+        InputOutput<Iterable<? extends I_ITEM>, O_ITEM> {
+
+    private Iterable<? extends I_ITEM> input;
+    private java.util.Map<String, String> options;
+    private Function<Iterable<? extends I_ITEM>, O_ITEM> function;
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java]: This should be annotated with "Required".

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/FlatMap.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ * an {@link Iterable} containing the resulting objects.
+ *
+ * @param <I_ITEM> The object type of the nested iterables
+ * @param <O_ITEM> the object type of the output iterable
+ */
+public class FlatMap<I_ITEM, O_ITEM> implements
+        InputOutput<Iterable<? extends Iterable<? extends I_ITEM>>, Iterable<? extends O_ITEM>> {
+    private Iterable<? extends Iterable<? extends I_ITEM>> input;
+    private Map<String, String> options;
+    private Function<Iterable<? extends I_ITEM>, O_ITEM> function;
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/FlatMap.java]: This should be annotated with "Required".

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/FlatMapHandler.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            throw new OperationException("Input cannot be null");
+        }
+
+        final Function<Iterable<? extends I_ITEM>, O_ITEM> function = operation.getFunction();
+
+        if (null == function) {
+            throw new OperationException("Function cannot be null");
+        }
+
+        return Streams.toStream(input).flatMap(function.andThen(Stream::of)).collect(toList());
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/FlatMapHandler.java]: We shouldn't collect into a List as that will bring all the results into memory. We should just process the iterable lazily.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * A {@code Map} is an {@link Operation} which applies a {@link Function} to map an
+ * input {@link Iterable} to an output object
+ *
+ * @param <I_ITEM> The object type of the input iterable
+ * @param <O_ITEM> The object type of the output object
+ */
+public class Map<I_ITEM, O_ITEM> implements
+        InputOutput<Iterable<? extends I_ITEM>, O_ITEM> {
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java]: You should also implement MultiInput<I_ITEM>

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Override
+    public TypeReference<O_ITEM> getOutputTypeReference() {
+        return TypeReferenceImpl.createExplicitT();
+    }
+
+    public static final class Builder<I_ITEM, O_ITEM> extends
+            Operation.BaseBuilder<Map<I_ITEM, O_ITEM>, Builder<I_ITEM, O_ITEM>> implements
+            InputOutput.Builder<Map<I_ITEM, O_ITEM>, Iterable<? extends I_ITEM>, O_ITEM, Builder<I_ITEM, O_ITEM>> {
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java]: You should also implement MultiInput.Builder

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/FlatMap.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * A {@code FlatMap} is an {@link Operation} which takes an {@link Iterable} of {@link Iterable}s,
+ * applies a function to map each of the nested {@link Iterable}s to an object, and returns
+ * an {@link Iterable} containing the resulting objects.
+ *
+ * @param <I_ITEM> The object type of the nested iterables
+ * @param <O_ITEM> the object type of the output iterable
+ */
+public class FlatMap<I_ITEM, O_ITEM> implements
+        InputOutput<Iterable<? extends Iterable<? extends I_ITEM>>, Iterable<? extends O_ITEM>> {
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/FlatMap.java]: You should also implement MultiInput<Iterable<? extends I_ITEM>>

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/FlatMap.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @Override
+    public TypeReference<Iterable<? extends O_ITEM>> getOutputTypeReference() {
+        return TypeReferenceImpl.createIterableT();
+    }
+
+    public static final class Builder<I_ITEM, O_ITEM> extends
+            Operation.BaseBuilder<FlatMap<I_ITEM, O_ITEM>, Builder<I_ITEM, O_ITEM>> implements
+            InputOutput.Builder<FlatMap<I_ITEM, O_ITEM>, Iterable<? extends Iterable<? extends I_ITEM>>, Iterable<? extends O_ITEM>, Builder<I_ITEM, O_ITEM>> {
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/FlatMap.java]: You should also implement MultiInput.Builder

File: pom.xml [hunk 10]
```diff
@@ -57,7 +57,7 @@
         <failsafe.version>2.18.1</failsafe.version>
         <slf4j.api.version>1.7.5</slf4j.api.version>
 
-        <koryphe.version>1.0.0</koryphe.version>
+        <koryphe.version>1.0.1-SNAPSHOT</koryphe.version>
```
Comment (p013570) [inline pom.xml]: This PR cannot be merged in until Koryphe 1.1.0 has been released. We cannot depend on SNAPSHOT versions.

**Comment by m607123** [comment]
Closing while changes made

File: core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/iterable/IterableUtil.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.commonutil.iterable;
+
+import java.util.Iterator;
+import java.util.function.Function;
+
+public class IterableUtil {
```
Comment (t616178) [inline core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/iterable/IterableUtil.java]: Is this meant to be a copy of the Koryphe class, or should we just be using that one instead?

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+ * @param <I> the type of the input object
+ * @param <O> the type of the output object
+ */
+public class Map<I, O> implements InputOutput<I, O> {
+    private I input;
+    private java.util.Map<String, String> options;
+    @Required
+    private Function<I, O> function;
+
+    public Map() {
```
Comment (t616178) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java]: Is this explicit constructor required by Jackson?

**Comment by t616178** [comment]
There are also some Checkstyle issues that need resolving.

**Comment by m607123** [comment]
Now depends on gchq/koryphe#60

**Pull request merged**